crates/events/src/extend.rs

```diff
@@ -666,6 +668,90 @@ where
     }
 }
 
+/// Implement the Display and FromStr traits for any serde type
+#[derive(Default, Clone, Serialize, Deserialize)]
+pub struct EventValue<T>(pub T);
+
+impl<T: Serialize> Display for EventValue<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let ser =
+            serde_json::to_string(&self.0).map_err(|_| std::fmt::Error)?;
+        write!(f, "{}", ser)
+    }
+}
+
+impl<T: for<'de> Deserialize<'de>> FromStr for EventValue<T> {
+    type Err = serde_json::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        serde_json::from_str(s).map(Self)
+    }
+}
+
+impl<T> From<T> for EventValue<T> {
+    fn from(t: T) -> Self {
+        Self(t)
+    }
+}
+
+/// A user account.
+#[derive(Debug, Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub enum UserAccount {
+    /// Internal chain address in Namada.
+    Internal(Address),
+    /// External chain address.
+    External(String),
+}
+
+impl fmt::Display for UserAccount {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Internal(addr) => write!(f, "internal-address/{addr}"),
+            Self::External(addr) => write!(f, "external-address/{addr}"),
+        }
+    }
+}
+
+impl FromStr for UserAccount {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.split_once('/') {
+            Some(("internal-address", addr)) => {
+                Ok(Self::Internal(Address::decode(addr).map_err(|err| {
+                    format!(
+                        "Unknown internal address balance change target \
+                         {s:?}: {err}"
+                    )
+                })?))
+            }
+            Some(("external-address", addr)) => {
+                Ok(Self::External(addr.to_owned()))
+            }
+            _ => Err(format!("Unknown balance change target {s:?}")),
+        }
+    }
+}
+
+impl serde::Serialize for UserAccount {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.collect_str(&self.to_string())
+    }
+}
+
+impl<'de> Deserialize<'de> for UserAccount {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = <String as Deserialize>::deserialize(deserializer)?;
+        FromStr::from_str(&s).map_err(serde::de::Error::custom)
+    }
+}
```

Comment:
replace with `namada::core::string_encoding::StringEncoded`, which uses `Display` and `FromStr` to encode some value with `serde`

Reply:
`StringEncoded<UserAccount>` that is

Reply:
@murisi can we try to address any further items related to this PR in a follow-up PR? This has been merged to draft now.

Reply:
Sure, understood. Thanks.